src/box/memtx_engine.cc

```diff
@@ -1073,6 +1055,20 @@ checkpoint_f(va_list ap)
 						    result.data, result.size);
 			if (rc != 0)
 				break;
+			struct errinj *inj = errinj(ERRINJ_SNAP_WRITE_TIMEOUT,
+						    ERRINJ_DOUBLE);
```

Comment:
Better just yield until the fiber is woken up or a timeout happens.

Reply:
Changed as discussed f2f:
- yield based on time instead of base of loop counter
- move delay to wal

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 609a72304..abb0b3d67 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -68,6 +68,9 @@
 /* sync snapshot every 16MB */
 #define SNAP_SYNC_INTERVAL	(1 << 24)
 
+/* How frequently checkpoint thread yields. */
+#define CHECKPOINT_YIELD_PERIOD 0.1f
+
 enum {
 	OBJSIZE_MIN = 16,
 	SLAB_SIZE = 16 * 1024 * 1024,
@@ -982,7 +985,6 @@ checkpoint_write_invalid_system_row(struct xlog *l)
 static int
 checkpoint_f(va_list ap)
 {
-	enum { YIELD_LOOPS = 1000 };
 	int rc = 0;
 	struct checkpoint *ckpt = va_arg(ap, struct checkpoint *);
 
@@ -1009,8 +1011,6 @@ checkpoint_f(va_list ap)
 	}
 
 	struct mh_i32_t *temp_space_ids;
-	unsigned count = 0;
-	unsigned yield_loops = YIELD_LOOPS;
 
 	say_info("saving snapshot `%s'", snap->filename);
 	ERROR_INJECT_WHILE(ERRINJ_SNAP_WRITE_DELAY, {
@@ -1040,6 +1040,7 @@ checkpoint_f(va_list ap)
 			rc = -1;
 			break;
 		}
+		double last_yield_time = ev_monotonic_now(loop());
 		while (true) {
 			RegionGuard region_guard(&fiber()->gc);
 			struct read_view_tuple result;
@@ -1055,15 +1056,12 @@ checkpoint_f(va_list ap)
 						    result.data, result.size);
 			if (rc != 0)
 				break;
-			struct errinj *inj = errinj(ERRINJ_SNAP_WRITE_TIMEOUT,
-						    ERRINJ_DOUBLE);
-			if (inj != NULL && inj->dparam > 0) {
-				thread_sleep(inj->dparam);
-				yield_loops = 1;
-			}
 			/* Yield to make thread cancellable. */
-			if (++count % yield_loops == 0)
+			if (ev_monotonic_now(loop()) - last_yield_time >
+			    CHECKPOINT_YIELD_PERIOD) {
+				last_yield_time = ev_monotonic_now(loop());
 				fiber_sleep(0);
+			}
 			if (fiber_is_cancelled()) {
 				diag_set(FiberIsCancelled);
 				rc = -1;
diff --git a/src/box/xlog.c b/src/box/xlog.c
index 5b7c550eb..9b768ee57 100644
--- a/src/box/xlog.c
+++ b/src/box/xlog.c
@@ -1230,6 +1230,9 @@ xlog_tx_write(struct xlog *log)
 	} else {
 		written = xlog_tx_write_plain(log);
 	}
+	/* In case we want to yield depending on how much time pass. */
+	if (loop())
+		ev_now_update(loop());
 	ERROR_INJECT(ERRINJ_WAL_WRITE, {
 		diag_set(ClientError, ER_INJECTION, "xlog write injection");
 		written = -1;
@@ -1357,6 +1360,13 @@ xlog_write_row(struct xlog *log, const struct xrow_header *packet)
 	    xlog_tx_write(log) < 0)
 		return -1;
 
+	struct errinj *inj = errinj(ERRINJ_WAL_WRITE_TIMEOUT, ERRINJ_DOUBLE);
+	if (inj != NULL && inj->dparam > 0) {
+		thread_sleep(inj->dparam);
+		if (loop())
+			ev_now_update(loop());
+	}
+
 	return row_size;
 }
 
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index a8ddba4ef..e35857030 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -134,7 +134,6 @@ struct errinj {
 	_(ERRINJ_SNAP_WRITE_INVALID_SYSTEM_ROW, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_SNAP_WRITE_MISSING_SPACE_ROW, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_SNAP_WRITE_UNKNOWN_ROW_TYPE, ERRINJ_BOOL, {.bparam = false}) \
-	_(ERRINJ_SNAP_WRITE_TIMEOUT, ERRINJ_DOUBLE, {.dparam = 0}) \
 	_(ERRINJ_SPACE_UPGRADE_DELAY, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_SWIM_FD_ONLY, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_TESTING, ERRINJ_BOOL, {.bparam = false}) \
@@ -186,6 +185,7 @@ struct errinj {
 	_(ERRINJ_WAL_WRITE_DISK, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_WAL_WRITE_EOF, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_WAL_WRITE_PARTIAL, ERRINJ_INT, {.iparam = -1}) \
+	_(ERRINJ_WAL_WRITE_TIMEOUT, ERRINJ_DOUBLE, {.dparam = 0}) \
 	_(ERRINJ_XLOG_GARBAGE, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_XLOG_META, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_XLOG_READ, ERRINJ_INT, {.iparam = -1}) \
diff --git a/test/app-luatest/shutdown_test.lua b/test/app-luatest/shutdown_test.lua
index 904357597..f7b7c1150 100644
--- a/test/app-luatest/shutdown_test.lua
+++ b/test/app-luatest/shutdown_test.lua
@@ -142,7 +142,7 @@ g_snap.test_shutdown_during_memtx_snapshot = function(cg)
         for i=1,10000 do
             box.space.test:insert{i}
         end
-        box.error.injection.set('ERRINJ_SNAP_WRITE_TIMEOUT', 0.1)
+        box.error.injection.set('ERRINJ_WAL_WRITE_TIMEOUT', 0.1)
         fiber.create(function()
             box.snapshot()
         end)
diff --git a/test/box/errinj.result b/test/box/errinj.result
index af5dcc6ce..e1950292c 100644
--- a/test/box/errinj.result
+++ b/test/box/errinj.result
@@ -106,7 +106,6 @@ evals
   - ERRINJ_SNAP_WRITE_DELAY: false
   - ERRINJ_SNAP_WRITE_INVALID_SYSTEM_ROW: false
   - ERRINJ_SNAP_WRITE_MISSING_SPACE_ROW: false
-  - ERRINJ_SNAP_WRITE_TIMEOUT: 0
   - ERRINJ_SNAP_WRITE_UNKNOWN_ROW_TYPE: false
   - ERRINJ_SPACE_UPGRADE_DELAY: false
   - ERRINJ_SWIM_FD_ONLY: false
@@ -159,6 +158,7 @@ evals
   - ERRINJ_WAL_WRITE_DISK: false
   - ERRINJ_WAL_WRITE_EOF: false
   - ERRINJ_WAL_WRITE_PARTIAL: -1
+  - ERRINJ_WAL_WRITE_TIMEOUT: 0
   - ERRINJ_XLOG_GARBAGE: false
   - ERRINJ_XLOG_META: false
   - ERRINJ_XLOG_READ: -1

```

</details>